### PR TITLE
lib/api: Save session & CSRF tokens to database, add option to stay logged in (fixes #9151)

### LINKED ISSUE #9151: Add "stay logged in" checkbox to login dialog
We should have an option to keep the cookie longer than session context.

### PR DESCRIPTION
**Pull request by calmh**
This adds a "token manager" which handles storing and checking expired tokens, used for both sessions and CSRF tokens. It removes the old, corresponding functionality for CSRFs which saved things in a file. The result is less crap in the state directory, and active login sessions now survive a Syncthing restart (this really annoyed me).

It also adds a boolean on login to create a longer-lived session cookie, which is now possible and useful. Thus we can remain logged in over browser restarts, which was also annoying... :)

Screenshot 2023-12-12 at 09 56 34

Best viewed with whitespace-insensitive diff, as a bunch of the auth functions became methods instead of closures which changed indentation.

### CODE REVIEW

File: lib/api/api_auth.go [hunk 1]
```diff
@@ -353,3 +368,124 @@ func iso88591ToUTF8(s []byte) []byte {
 	}
 	return []byte(string(runes))
 }
+
+type tokenManager struct {
```
Comment (imsodin) [inline lib/api/api_auth.go:372]: Optional/opinionated: I'd move the tokenManager code to it's own file - it's a non-trivial amount of code and used both here and in the csrf file/code.

**Review by imsodin — APPROVED** [review]
Looks good - one optional suggestion.

I was initially quite dubious about the value of the proto map, as it feels a bit superfluous when we could put it directly into the db with subkeys. However handling multiple db values probably amounts to more boilerplate than this :)

File: lib/api/api_auth.go [hunk 2]
```diff
@@ -353,3 +368,124 @@ func iso88591ToUTF8(s []byte) []byte {
...
+}
+
+// New creates a new token and returns it.
+func (m *tokenManager) New() string {
+	token := rand.String(randomTokenLength)
+
+	m.mut.Lock()
+	defer m.mut.Unlock()
+
+	m.tokens.Tokens[token] = m.timeNow().Add(m.lifetime).UnixNano()
```
[the remark this answers is not in the document]
Comment (calmh) [reply lib/api/api_auth.go:431]: I think initially it was easier when I did testing with actual clock time, as sorting and expiry didn't work well with just one second granularity, and then I figured it never hurts.

**Comment by calmh** [comment]
> However handling multiple db values probably amounts to more boilerplate than this 

Kinda sorta maybe because then we need a whole namespace so we can iterate for cleanup etc. This seemed simpler / more self contained. 🤷

**Pull request merged**